File: ALittle/Module/ALittleIDE/Other/GameLibrary/SEngine/JSScript/CacheData/CacheDataSet.js
{
if (typeof ALittle === "undefined") ALittle = {};

ALittle.RegStruct(-1164907202, "ALittle.CacheData", {
name : "ALittle.CacheData", ns_name : "ALittle", rl_name : "CacheData", hash_code : -1164907202,
name_list : ["id"],
type_list : ["int"],
option_map : {primary : "id"}
})

ALittle.CacheDataSet = JavaScript.Class(undefined, {
	Ctor : function(submit_interval_ms, cache_interval_ms) {
		this._submit_interval_ms = submit_interval_ms;
		this._cache_interval_ms = cache_interval_ms;
		this._data_map = new Map();
		this._loading_map = new Map();
		this._timer_map = new Map();
		this._dirty_map = new Map();
	},
	Init : function() {
		return new Promise((async function(___COROUTINE, ___) {
			let rflt = this.__class.__element[0];
			this._primary = rflt.option_map["primary"];
			if (this._primary === undefined) {
				___COROUTINE(rflt.name + "没有定义主键"); return;
			}
			let error = await A_MysqlSystem.CreateIfNotExit(this.__class.__element[0]);
			if (error !== undefined) {
				___COROUTINE(error); return;
			}
			if (this._submit_timer !== undefined) {
				A_LoopSystem.RemoveTimer(this._submit_timer);
			}
			this._submit_timer = A_LoopSystem.AddTimer(this._submit_interval_ms, this.Submit.bind(this, true));
			___COROUTINE(undefined); return;
			___COROUTINE();
		}).bind(this));
	},
	Release : function() {
		if (this._submit_timer !== undefined) {
			A_LoopSystem.RemoveTimer(this._submit_timer);
			this._submit_timer = undefined;
		}
		this.Submit(false);
		this._data_map = new Map();
		this._loading_map = new Map();
		for (let [id, timer] of this._timer_map) {
			if (timer === undefined) continue;
			A_LoopSystem.RemoveTimer(timer);
		}
		this._timer_map = new Map();
		this._dirty_map = new Map();
		this._release = true;
		let rflt = this.__class.__element[0];
		ALittle.Log(rflt.name + "操作完毕");
	},
	Submit : async function(loop) {
		this._submit_timer = undefined;
		let data_map = this._data_map;
		let dirty_map = this._dirty_map;
		for (let [id, _] of dirty_map) {
			if (_ === undefined) continue;
			let data = data_map.get(id);
			if (data !== undefined) {
				let error = await A_MysqlSystem.UpdateOne(this.__class.__element[0], data, this._primary, data.id, data.id);
				if (error !== undefined) {
					ALittle.Error(error);
				}
			}
		}
		this._dirty_map = new Map();
		if (loop) {
			this._submit_timer = A_LoopSystem.AddTimer(this._submit_interval_ms, this.Submit.bind(this, true));
		}
	},
	GetDataAndDirty : function(id) {
		return new Promise((async function(___COROUTINE, ___) {
			let data = await this.GetData(id);
			if (data !== undefined) {
				this._dirty_map.set(id, true);
			}
			___COROUTINE(data); return;
			___COROUTINE();
		}).bind(this));
	},
	GetData : function(id) {
		return new Promise((async function(___COROUTINE, ___) {
			if (this._release) {
				___COROUTINE(undefined); return;
			}
			let data = this._data_map.get(id);
			if (data !== undefined) {
				___COROUTINE(data); return;
			}
			if (this._loading_map.get(id)) {
				await A_MysqlSystem.ExecuteEmpty(id);
				if (this._release) {
					___COROUTINE(undefined); return;
				}
				___COROUTINE(this._data_map.get(id)); return;
			}
			this._loading_map.set(id, true);
			let [error, new_data] = await A_MysqlSystem.SelectOneFromByKey(this.__class.__element[0], this._primary, id, id);
			if (error !== undefined) {
				ALittle.Error(error);
			}
			if (this._release) {
				___COROUTINE(undefined); return;
			}
			this._loading_map.delete(id);
			if (new_data === undefined) {
				___COROUTINE(undefined); return;
			}
			this._data_map.set(id, new_data);
			if (this._timer_map.get(id) !== undefined) {
				A_LoopSystem.RemoveTimer(this._timer_map.get(id));
			}
			this._timer_map.set(id, A_LoopSystem.AddTimer(this._cache_interval_ms, this.RemoveCache.bind(this, id)));
			___COROUTINE(new_data); return;
			___COROUTINE();
		}).bind(this));
	},
	RemoveCache : function(id) {
		if (this._release) {
			return;
		}
		this._timer_map.delete(id);
		let data = this._data_map.get(id);
		if (data === undefined) {
			return;
		}
		this._data_map.delete(id);
		if (this._dirty_map.get(id)) {
			this._dirty_map.delete(data.id);
			this.SubmitData(data);
		}
	},
	SubmitData : async function(data) {
		let error = await A_MysqlSystem.UpdateOne(this.__class.__element[0], data, this._primary, data.id, data.id);
		if (error !== undefined) {
			ALittle.Error(error);
		}
	},
	CreateData : function(data) {
		return new Promise((async function(___COROUTINE, ___) {
			if (this._release) {
				___COROUTINE("数据集已经被释放"); return;
			}
			if (this._loading_map.get(data.id)) {
				await A_MysqlSystem.ExecuteEmpty(data.id);
				if (this._release) {
					___COROUTINE("数据集已经被释放"); return;
				}
			}
			if (this._data_map.get(data.id) !== undefined) {
				___COROUTINE("数据已存在"); return;
			}
			this._data_map.set(data.id, data);
			if (this._timer_map.get(data.id) !== undefined) {
				A_LoopSystem.RemoveTimer(this._timer_map.get(data.id));
			}
			this._timer_map.set(data.id, A_LoopSystem.AddTimer(this._cache_interval_ms, this.RemoveCache.bind(this, data.id)));
			let error = await A_MysqlSystem.InsertInto(this.__class.__element[0], data, undefined, data.id);
			if (error !== undefined) {
				ALittle.Error(error);
			}
			___COROUTINE(undefined); return;
			___COROUTINE();
		}).bind(this));
	},
	DeleteData : function(id) {
		return new Promise((async function(___COROUTINE, ___) {
			if (this._release) {
				___COROUTINE("数据集已经被释放"); return;
			}
			if (this._loading_map.get(id)) {
				await A_MysqlSystem.ExecuteEmpty(id);
				if (this._release) {
					___COROUTINE("数据集已经被释放"); return;
				}
			}
			this._dirty_map.delete(id);
			let timer = this._timer_map.get(id);
			if (timer !== undefined) {
				A_LoopSystem.RemoveTimer(timer);
				this._timer_map.delete(id);
			}
			this._data_map.delete(id);
			let error = await A_MysqlSystem.DeleteFromByKey(this.__class.__element[0], this._primary, id, id);
			if (error !== undefined) {
				ALittle.Error(error);
			}
			___COROUTINE(undefined); return;
			___COROUTINE();
		}).bind(this));
	},
}, "ALittle.CacheDataSet");

}

File: ALittle/Module/ALittleIDE/Other/GameLibrary/SEngine/Script/CacheData/CacheDataSet.lua

module("ALittle", package.seeall)

local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs
local ___coroutine = coroutine

RegStruct(-1164907202, "ALittle.CacheData", {
name = "ALittle.CacheData", ns_name = "ALittle", rl_name = "CacheData", hash_code = -1164907202,
name_list = {"id"},
type_list = {"int"},
option_map = {primary="id"}
})

CacheDataSet = Class(nil, "ALittle.CacheDataSet")

function CacheDataSet:Ctor(submit_interval_ms, cache_interval_ms)
	___rawset(self, "_submit_interval_ms", submit_interval_ms)
	___rawset(self, "_cache_interval_ms", cache_interval_ms)
	___rawset(self, "_data_map", {})
	___rawset(self, "_loading_map", {})
	___rawset(self, "_timer_map", {})
	___rawset(self, "_dirty_map", {})
end

function CacheDataSet:Init()
	local rflt = self.__class.__element[1]
	self._primary = rflt.option_map["primary"]
	if self._primary == nil then
		return rflt.name .. "没有定义主键"
	end
	local error = A_MysqlSystem:CreateIfNotExit(self.__class.__element[1])
	if error ~= nil then
		return error
	end
	if self._submit_timer ~= nil then
		A_LoopSystem:RemoveTimer(self._submit_timer)
	end
	self._submit_timer = A_LoopSystem:AddTimer(self._submit_interval_ms, Bind(self.Submit, self, true))
	return nil
end

function CacheDataSet:Release()
	if self._submit_timer ~= nil then
		A_LoopSystem:RemoveTimer(self._submit_timer)
		self._submit_timer = nil
	end
	self:Submit(false)
	self._data_map = {}
	self._loading_map = {}
	for id, timer in ___pairs(self._timer_map) do
		A_LoopSystem:RemoveTimer(timer)
	end
	self._timer_map = {}
	self._dirty_map = {}
	self._release = true
	local rflt = self.__class.__element[1]
	Log(rflt.name .. "操作完毕")
end

function CacheDataSet:Submit(loop)
	self._submit_timer = nil
	local data_map = self._data_map
	local dirty_map = self._dirty_map
	for id, _ in ___pairs(dirty_map) do
		local data = data_map[id]
		if data ~= nil then
			local error = A_MysqlSystem:UpdateOne(self.__class.__element[1], data, self._primary, data.id, data.id)
			if error ~= nil then
				Error(error)
			end
		end
	end
	self._dirty_map = {}
	if loop then
		self._submit_timer = A_LoopSystem:AddTimer(self._submit_interval_ms, Bind(self.Submit, self, true))
	end
end
CacheDataSet.Submit = CoWrap(CacheDataSet.Submit)

function CacheDataSet:GetDataAndDirty(id)
	local data = self:GetData(id)
	if data ~= nil then
		self._dirty_map[id] = true
	end
	return data
end

function CacheDataSet:GetData(id)
	if self._release then
		return nil
	end
	local data = self._data_map[id]
	if data ~= nil then
		return data
	end
	if self._loading_map[id] then
		A_MysqlSystem:ExecuteEmpty(id)
		if self._release then
			return nil
		end
		return self._data_map[id]
	end
	self._loading_map[id] = true
	local error, new_data = A_MysqlSystem:SelectOneFromByKey(self.__class.__element[1], self._primary, id, id)
	if error ~= nil then
		Error(error)
	end
	if self._release then
		return nil
	end
	self._loading_map[id] = nil
	if new_data == nil then
		return nil
	end
	self._data_map[id] = new_data
	if self._timer_map[id] ~= nil then
		A_LoopSystem:RemoveTimer(self._timer_map[id])
	end
	self._timer_map[id] = A_LoopSystem:AddTimer(self._cache_interval_ms, Bind(self.RemoveCache, self, id))
	return new_data
end

function CacheDataSet:RemoveCache(id)
	if self._release then
		return
	end
	self._timer_map[id] = nil
	local data = self._data_map[id]
	if data == nil then
		return
	end
	self._data_map[id] = nil
	if self._dirty_map[id] then
		self._dirty_map[data.id] = nil
		self:SubmitData(data)
	end
end

function CacheDataSet:SubmitData(data)
	local error = A_MysqlSystem:UpdateOne(self.__class.__element[1], data, self._primary, data.id, data.id)
	if error ~= nil then
		Error(error)
	end
end
CacheDataSet.SubmitData = CoWrap(CacheDataSet.SubmitData)

function CacheDataSet:CreateData(data)
	if self._release then
		return "数据集已经被释放"
	end
	if self._loading_map[data.id] then
		A_MysqlSystem:ExecuteEmpty(data.id)
		if self._release then
			return "数据集已经被释放"
		end
	end
	if self._data_map[data.id] ~= nil then
		return "数据已存在"
	end
	self._data_map[data.id] = data
	if self._timer_map[data.id] ~= nil then
		A_LoopSystem:RemoveTimer(self._timer_map[data.id])
	end
	self._timer_map[data.id] = A_LoopSystem:AddTimer(self._cache_interval_ms, Bind(self.RemoveCache, self, data.id))
	local error = A_MysqlSystem:InsertInto(self.__class.__element[1], data, nil, data.id)
	if error ~= nil then
		Error(error)
	end
	return nil
end

function CacheDataSet:DeleteData(id)
	if self._release then
		return "数据集已经被释放"
	end
	if self._loading_map[id] then
		A_MysqlSystem:ExecuteEmpty(id)
		if self._release then
			return "数据集已经被释放"
		end
	end
	self._dirty_map[id] = nil
	local timer = self._timer_map[id]
	if timer ~= nil then
		A_LoopSystem:RemoveTimer(timer)
		self._timer_map[id] = nil
	end
	self._data_map[id] = nil
	local error = A_MysqlSystem:DeleteFromByKey(self.__class.__element[1], self._primary, id, id)
	if error ~= nil then
		Error(error)
	end
	return nil
end
